File: slp/backend-dotnet/Features/Admin/AdminController.cs
using backend_dotnet.Features.Admin;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;

namespace backend_dotnet.Features.Admin;

[Authorize(Roles = "admin")]
[ApiController]
[Route("api/admin")]
public class AdminController : ControllerBase
{
    private readonly IAdminService _adminService;

    public AdminController(IAdminService adminService)
    {
        _adminService = adminService;
    }

    // Users
    [HttpGet("users")]
    public async Task<IActionResult> GetUsers(
        [FromQuery] string? search = null,
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 20)
    {
        page = page < 1 ? 1 : page;
        pageSize = pageSize < 1 ? 20 : (pageSize > 100 ? 100 : pageSize);
        var result = await _adminService.GetAllUsersAsync(search, page, pageSize);
        return Ok(result);
    }

    [HttpPost("users/{id}/ban")]
    public async Task<IActionResult> BanUser(int id)
    {
        var adminId = int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
        var banned = await _adminService.BanUserAsync(adminId, id);
        if (!banned) return BadRequest("Cannot ban admin or user not found");
        return Ok();
    }

    [HttpPost("users/{id}/unban")]
    public async Task<IActionResult> UnbanUser(int id)
    {
        var adminId = int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
        var unbanned = await _adminService.UnbanUserAsync(adminId, id);
        if (!unbanned) return NotFound();
        return Ok();
    }

    // Quizzes
    [HttpGet("quizzes")]
    public async Task<IActionResult> GetQuizzes(
        [FromQuery] string? search = null,
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 20)
    {
        page = page < 1 ? 1 : page;
        pageSize = pageSize < 1 ? 20 : (pageSize > 100 ? 100 : pageSize);
        var result = await _adminService.GetAllQuizzesAsync(search, page, pageSize);
        return Ok(result);
    }

    [HttpPost("quizzes/{id}/disable")]
    public async Task<IActionResult> DisableQuiz(int id)
    {
        var adminId = int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
        var disabled = await _adminService.DisableQuizAsync(adminId, id);
        if (!disabled) return NotFound();
        return Ok();
    }

    [HttpPost("quizzes/{id}/enable")]
    public async Task<IActionResult> EnableQuiz(int id)
    {
        var adminId = int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
        var enabled = await _adminService.EnableQuizAsync(adminId, id);
        if (!enabled) return NotFound();
        return Ok();
    }

    // Comments
    [HttpGet("comments")]
    public async Task<IActionResult> GetComments(
        [FromQuery] bool includeDeleted = false,
        [FromQuery] string? search = null,
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 20)
    {
        page = page < 1 ? 1 : page;
        pageSize = pageSize < 1 ? 20 : (pageSize > 100 ? 100 : pageSize);
        var result = await _adminService.GetAllCommentsAsync(includeDeleted, search, page, pageSize);
        return Ok(result);
    }

    [HttpDelete("comments/{id}")]
    public async Task<IActionResult> DeleteComment(int id)
    {
        var adminId = int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
        var deleted = await _adminService.DeleteCommentAsync(adminId, id);
        if (!deleted) return NotFound();
        return NoContent();
    }

    [HttpPost("comments/{id}/restore")]
    public async Task<IActionResult> RestoreComment(int id)
    {
        var adminId = int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
        var restored = await _adminService.RestoreCommentAsync(adminId, id);
        if (!restored) return NotFound();
        return Ok();
    }

    // Logs
    [HttpGet("logs")]
    public async Task<IActionResult> GetLogs(
        [FromQuery] AdminLogFilterDto filter,
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 20)
    {
        page = page < 1 ? 1 : page;
        pageSize = pageSize < 1 ? 20 : (pageSize > 100 ? 100 : pageSize);
        var result = await _adminService.GetRecentLogsAsync(filter, page, pageSize);
        return Ok(result);
    }
}

File: slp/backend-nestjs/src/modules/admin/admin.controller.ts
import {
  Controller,
  Get,
  Post,
  Delete,
  Param,
  Query,
  UseGuards,
  Request,
  HttpCode,
  HttpStatus,
} from "@nestjs/common";
import { AdminService } from "./admin.service";
import { SessionGuard } from "../session/session.guard";
import { RolesGuard } from "../../common/guards/roles.guard";
import { Roles } from "../../common/decorators/roles.decorator";
import { UserDto } from "./dto/user.dto";
import { QuizAdminDto } from "./dto/quiz-admin.dto";
import { CommentAdminDto } from "./dto/comment-admin.dto";
import { AdminLogDto } from "./dto/admin-log.dto";
import { AdminLogFilterDto } from "./dto/admin-log-filter.dto";
import { PaginatedResult } from "../../helpers/pagination.helper";

@Controller("api/admin")
@UseGuards(SessionGuard, RolesGuard)
@Roles("admin")
export class AdminController {
  constructor(private readonly adminService: AdminService) {}

  // Users
  @Get("users")
  async getUsers(
    @Query("search") search?: string,
    @Query("page") page = 1,
    @Query("pageSize") pageSize = 20,
  ): Promise<PaginatedResult<UserDto>> {
    page = page < 1 ? 1 : page;
    pageSize = pageSize < 1 ? 20 : pageSize > 100 ? 100 : pageSize;
    return this.adminService.getAllUsers(search, page, pageSize);
  }

  @Post("users/:id/ban")
  @HttpCode(HttpStatus.OK)
  async banUser(@Request() req, @Param("id") id: string): Promise<void> {
    await this.adminService.banUser(req.user.id, parseInt(id, 10));
  }

  @Post("users/:id/unban")
  @HttpCode(HttpStatus.OK)
  async unbanUser(@Request() req, @Param("id") id: string): Promise<void> {
    await this.adminService.unbanUser(req.user.id, parseInt(id, 10));
  }

  // Quizzes
  @Get("quizzes")
  async getQuizzes(
    @Query("search") search?: string,
    @Query("page") page = 1,
    @Query("pageSize") pageSize = 20,
  ): Promise<PaginatedResult<QuizAdminDto>> {
    page = page < 1 ? 1 : page;
    pageSize = pageSize < 1 ? 20 : pageSize > 100 ? 100 : pageSize;
    return this.adminService.getAllQuizzes(search, page, pageSize);
  }

  @Post("quizzes/:id/disable")
  @HttpCode(HttpStatus.OK)
  async disableQuiz(@Request() req, @Param("id") id: string): Promise<void> {
    await this.adminService.disableQuiz(req.user.id, parseInt(id, 10));
  }

  @Post("quizzes/:id/enable")
  @HttpCode(HttpStatus.OK)
  async enableQuiz(@Request() req, @Param("id") id: string): Promise<void> {
    await this.adminService.enableQuiz(req.user.id, parseInt(id, 10));
  }

  // Comments
  @Get("comments")
  async getComments(
    @Query("includeDeleted") includeDeleted = "false",
    @Query("search") search?: string,
    @Query("page") page = 1,
    @Query("pageSize") pageSize = 20,
  ): Promise<PaginatedResult<CommentAdminDto>> {
    page = page < 1 ? 1 : page;
    pageSize = pageSize < 1 ? 20 : pageSize > 100 ? 100 : pageSize;
    return this.adminService.getAllComments(
      includeDeleted === "true",
      search,
      page,
      pageSize,
    );
  }

  @Delete("comments/:id")
  @HttpCode(HttpStatus.NO_CONTENT)
  async deleteComment(@Request() req, @Param("id") id: string): Promise<void> {
    await this.adminService.deleteComment(req.user.id, parseInt(id, 10));
  }

  @Post("comments/:id/restore")
  @HttpCode(HttpStatus.OK)
  async restoreComment(@Request() req, @Param("id") id: string): Promise<void> {
    await this.adminService.restoreComment(req.user.id, parseInt(id, 10));
  }

  // Logs
  @Get("logs")
  async getLogs(
    @Query() filter: AdminLogFilterDto,
    @Query("page") page = 1,
    @Query("pageSize") pageSize = 20,
  ): Promise<PaginatedResult<AdminLogDto>> {
    page = page < 1 ? 1 : page;
    pageSize = pageSize < 1 ? 20 : pageSize > 100 ? 100 : pageSize;
    return this.adminService.getRecentLogs(filter, page, pageSize);
  }
}
